internal/eventstore/v3/search.go

```diff
@@ -0,0 +1,306 @@
+package eventstore
+
+import (
+	"context"
+	"database/sql"
+	_ "embed"
+	"strconv"
+	"strings"
+
+	"github.com/jackc/pgx/v5/pgtype"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/eventstore"
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+)
+
+func (es *Eventstore) Search(ctx context.Context, conditions ...map[eventstore.SearchFieldType]any) (result []*eventstore.SearchResult, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer span.EndWithError(err)
+
+	var builder strings.Builder
+	args := buildSearchStatement(ctx, &builder, conditions...)
+
+	err = es.client.QueryContext(
+		ctx,
+		func(rows *sql.Rows) error {
+			for rows.Next() {
+				var (
+					res         eventstore.SearchResult
+					textValue   sql.Null[string]
+					numberValue pgtype.Numeric
+				)
+
+				err = rows.Scan(
+					&res.Aggregate.InstanceID,
+					&res.Aggregate.ResourceOwner,
+					&res.Aggregate.Type,
+					&res.Aggregate.ID,
+					&res.Object.Type,
+					&res.Object.ID,
+					&res.Object.Revision,
+					&res.FieldName,
+					&textValue,
+					&numberValue,
+				)
+				if err != nil {
+					return err
+				}
+
+				if numberValue.Valid {
+					value, err := numberValue.MarshalJSON()
+					if err != nil {
+						return err
+					}
+					res.Value = value
+					res.ValueType = eventstore.SearchValueTypeNumeric
+				} else if textValue.Valid {
+					res.Value = []byte(textValue.V)
+					res.ValueType = eventstore.SearchValueTypeString
+				}
+				result = append(result, &res)
+			}
+			return nil
+		},
+		builder.String(),
+		args...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+const searchQueryPrefix = `SELECT instance_id, resource_owner, aggregate_type, aggregate_id, object_type, object_id, object_revision, field_name, text_value, number_value FROM eventstore.search WHERE instance_id = $1`
+
+func buildSearchStatement(ctx context.Context, builder *strings.Builder, conditions ...map[eventstore.SearchFieldType]any) []any {
+	args := make([]any, 0, len(conditions)*4+1)
+	args = append(args, authz.GetInstance(ctx).InstanceID())
+
+	builder.WriteString(searchQueryPrefix)
+
+	if len(conditions) == 0 {
+		return args
+	}
+
+	builder.WriteString(" AND (")
+
+	for i, condition := range conditions {
+		if i > 0 {
+			builder.WriteString(" OR ")
+		}
+		builder.WriteRune('(')
+		args = append(args, buildSearchCondition(builder, len(args)+1, condition)...)
+		builder.WriteRune(')')
+	}
+	builder.WriteRune(')')
+
+	return args
+}
+
+func buildSearchCondition(builder *strings.Builder, index int, conditions map[eventstore.SearchFieldType]any) []any {
+	args := make([]any, 0, len(conditions))
+
+	for field, value := range conditions {
+		if len(args) > 0 {
+			builder.WriteString(" AND ")
+		}
+		builder.WriteString(searchFieldNameByType(field))
+		builder.WriteString(" = $")
+		builder.WriteString(strconv.Itoa(index + len(args)))
+		args = append(args, value)
+	}
```

Comment:
Go maps are not sorted. So this will result in a different looking query on each execution.

1. We won't benefit from the implicit prepared statements optimization pgx is doing.
2. The order of search term won't be in the order of the defined index. Yes, the query optimizer can take care of this, but will that still work well with the nested `AND` and `OR`s we are doing here.
3. Repeated executions of the same query helps the optimizer. Shuffling search terms around might work counter productive
4. CloudSQL dashboard and sampling won't consider the different querys.

**If you insist** on using a map, I would suggest making it sorted with a ad-hoc slice:


```suggestion
	order := make([]eventstore.SearchFieldType, 0, len(conditions))
	for field := range conditions {
		order = append(order, field)
	}
	slices.Sort(order)

	for _, field := range order {
		if len(args) > 0 {
			builder.WriteString(" AND ")
		}
		builder.WriteString(searchFieldNameByType(field))
		builder.WriteString(" = $")
		builder.WriteString(strconv.Itoa(index + len(args)))
		args = append(args, conditions[field])
	}
```

**Alternativly**, does it not make sense to use a `struct` type instead of a `map` with enumerated keys. It would mean a bit more verbosity in the query builder here, but makes the life of package consumers a lot easier. Or perhaps using an interface `SearchCondition` with different struct implementations for the different supported indexed search operations.

Reply:
Or a slice of struct will be a intermediated solution, to maintain order, but still be flexible.

```go
type SearchCondition struct {
    Field SearchFieldType
    Value any
}

type Searcher interface {
	// Search predefined results
	// the instance is used from ctx
	Search(ctx context.Context, conditions ...[]SearchCondition) (result []*SearchResult, err error)
}
```

However, do we want to be flexible here? The only reason why we need a search table is to be able to do indexed searched on payloads. Using the search table in a way that is not covered by its indexes would beat its purpose. So I'm still in favor of the previous proposal of using a fixed struct for search conditions.